rest/server_context_test.go

```diff
@@ -522,3 +524,113 @@ func TestUseTLSServer(t *testing.T) {
 		})
 	}
 }
+
+func TestLogFlush(t *testing.T) {
+	testCases := []struct {
+		Name                 string
+		ExpectedLogFileCount int
+		EnableFunc           func(config StartupConfig) StartupConfig
+	}{
+		{
+			"Default",
+			4,
+			func(config StartupConfig) StartupConfig {
+				return config
+			},
+		},
+		{
+			"Add trace",
+			5,
+			func(config StartupConfig) StartupConfig {
+				config.Logging.Trace = &base.FileLoggerConfig{
+					Enabled: base.BoolPtr(true),
+				}
+				return config
+			},
+		},
+		{
+			"Add debug and trace",
+			6,
+			func(config StartupConfig) StartupConfig {
+				config.Logging.Debug = &base.FileLoggerConfig{
+					Enabled: base.BoolPtr(true),
+				}
+				config.Logging.Trace = &base.FileLoggerConfig{
+					Enabled: base.BoolPtr(true),
+				}
+				return config
+			},
+		},
+		{
+			"Disable error",
+			3,
+			func(config StartupConfig) StartupConfig {
+				config.Logging.Error = &base.FileLoggerConfig{
+					Enabled: base.BoolPtr(false),
+				}
+				return config
+			},
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.Name, func(t *testing.T) {
+			defer base.SetUpTestLogging(base.LevelInfo, base.KeyAll)()
+
+			// Setup memory logging
+			base.InitializeMemoryLoggers()
+
+			// Log some stuff (which will go into the memory loggers)
+			base.Errorf("error")
+			base.Warnf("warn")
+			base.Infof(base.KeyAll, "info")
+			base.Debugf(base.KeyAll, "debug")
+			base.Tracef(base.KeyAll, "trace")
+			base.RecordStats("{}")
+
+			// Add temp dir to save log files to
+			tempPath, err := ioutil.TempDir("", "logs"+testCase.Name)
+			require.NoError(t, err)
+
+			config := DefaultStartupConfig(tempPath)
+			config = testCase.EnableFunc(config)
+
+			// Setup logging
+			err = config.SetupAndValidateLogging()
+			assert.NoError(t, err)
+
+			// Flush memory loggers
+			base.FlushLoggerBuffers()
+
+			// Add a minor sleep to allow the files to complete writing
+			time.Sleep(1 * time.Second)
```

Comment:
Do we have to have a fixed sleep if we're checking in a retry loop? Maybe just increase the sleeper func's iterations or wait time.

Reply:
I think this should still be kept to stay safe.

Imagine in the default case we're waiting for 4 log files. If we hit the sleeper iteration and for whatever reason 'trace' was about to write but didn't quite get there (which I did see once in my tests) then we'd pass. 

Sleep gives an extra second for those writes to give us a little extra confidence any log files have written. Probably means we don't really need to have the sleeper too but will leave.

Reply:
I'd have expected the FlushLoggerBuffers call to be synchronous as long as we don't have collation buffers enabled (which we don't appear to for the memory loggers) but maybe that assumption is wrong

Reply:
I think it may well be collation buffers, I can turn them off. Why do you think we don't have them enabled?